File: Delphi/SoapSecurity/certlistfrm.pas
unit certlistfrm;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, clCertificate, clCertificateStore;

type
  TCertListForm = class(TForm)
    lbList: TListBox;
    btnInstall: TButton;
    btnClose: TButton;
    procedure btnInstallClick(Sender: TObject);
    procedure btnCloseClick(Sender: TObject);
  private
    FCertificates: TclCertificateStore;
    procedure FillList;
  public
    class procedure ShowCertificates(AList: TclCertificateStore);
  end;

var
  CertListForm: TCertListForm;

implementation

{$R *.dfm}

procedure TCertListForm.btnInstallClick(Sender: TObject);
begin
  if lbList.ItemIndex > -1 then
  begin
    FCertificates.Install(FCertificates.Items[lbList.ItemIndex]);
    ShowMessage('The message certificate has been installed.');
  end;
end;

class procedure TCertListForm.ShowCertificates(AList: TclCertificateStore);
var
  dlg: TCertListForm;
begin
  dlg := TCertListForm.Create(nil);
  try
    dlg.FCertificates := AList;
    dlg.FillList();
    dlg.ShowModal();
  finally
    dlg.Free();
  end;
end;

procedure TCertListForm.btnCloseClick(Sender: TObject);
begin
  Close();
end;

procedure TCertListForm.FillList;
var
  i: Integer;
  s: string;
begin
  lbList.Items.Clear();
  for i := 0 to FCertificates.Items.Count - 1 do
  begin
    s := FCertificates.Items[i].IssuedTo;
    if (FCertificates.Items[i].Email <> '') then
    begin
      s := s + ' (' + FCertificates.Items[i].Email + ')';
    end;
    lbList.Items.Add(s);
  end;
end;

end.

File: CBuilder/SMime/certlistfrm.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "certlistfrm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TCertListForm *CertListForm;
//---------------------------------------------------------------------------
__fastcall TCertListForm::TCertListForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TCertListForm::FillList()
{
  UnicodeString s;
  lbList->Items->Clear();
  for(int i = 0; i < FCertificates->Items->Count; i++)
  {
    s = FCertificates->Items->Items[i]->IssuedTo;
    if(FCertificates->Items->Items[i]->Email != "")
      s += " (" + FCertificates->Items->Items[i]->Email + ")";
    lbList->Items->Add(s);
  }
}
//---------------------------------------------------------------------------
void __fastcall TCertListForm::ShowCertificates(TclCertificateStore *AList)
{
  TCertListForm *dlg = new TCertListForm(NULL);
  __try
  {
    dlg->FCertificates = AList;
    dlg->FillList();
    dlg->ShowModal();
  }
  __finally
  {
    delete dlg;
  }
}
//---------------------------------------------------------------------------

void __fastcall TCertListForm::btnInstallClick(TObject *Sender)
{
  if(lbList->ItemIndex > -1)
  {
    FCertificates->Install(FCertificates->Items->Items[lbList->ItemIndex]);
    ShowMessage("The message certificate has been installed.");
  }
}
//---------------------------------------------------------------------------

void __fastcall TCertListForm::btnCloseClick(TObject *Sender)
{
  Close();
}
//---------------------------------------------------------------------------

